### PR TITLE
[ROCm][CI] Loosen block-FP8 fused MoE test tolerance for large-K shapes

### PR DESCRIPTION
**Pull request by stefankoncarevic**
## Purpose

`tests/kernels/moe/test_block_fp8.py::test_w8a8_block_fp8_fused_moe` compares two
Triton block-FP8 fused-MoE kernels (`fused_experts` and `modular_triton_fused_moe`)
against a native torch reference. On ROCm/gfx950 (MI355X) the large-K (`K=7168`),
large-N (`N >= 1024`) DeepSeek-style shapes failed the comparison at the base
tolerance.

Digging in, the divergence turned out to be **reference artifacts, not kernel
error** — so instead of loosening the tolerance, this PR fixes the reference and
keeps the tight base tolerance:

1. **fp8 quant boundary flips.** The reference quantized activations with a plain
   torch path whose fp8 rounding differed from the production
   `per_token_group_quant_fp8` op the kernels actually use (~0.06-0.10% of elements
   flip to the adjacent fp8 code). Over `K=7168` these accumulate. Fixed by
   quantizing the reference with the **same production op** (bit-identical quant),
   with the matmul kept in fp32.
2. **SiLU precision mismatch.** `modular_triton_fused_moe` fuses SiLU+quant in fp32
   (`silu_and_mul_per_block_quant`), while `fused_experts` and the reference round
   the SiLU output to bf16 first. On large K/N this ~1-ULP gap pushed the modular
   output past the base tolerance. Fixed by validating the modular output against a
   **fp32-SiLU reference** on ROCm large-K/N, again keeping the tight tolerance.

The tolerance itself is **unchanged and identical on all platforms**
(`atol=rtol=0.035`, or `0.039` for `M >= 8192`) — there is no per-architecture or
widened tolerance. The only platform branch (`is_rocm() and K >= 4096 and N >= 1024`)
swaps in the more faithful fp32-SiLU reference; it does not touch the tolerance. 
Test-only change, no production code is modified.

### History (how this evolved)

An earlier revision widened the ROCm tolerance to `0.08` for these shapes. Per
review feedback that was reworked: the widening was removed and replaced by the
reference fixes above, so the strict base tolerance is retained.

- `019d0786b` — (superseded) scoped ROCm large-K/N tolerance widening.
- `8ae3dea10` — align the test reference quant with the production kernel.
- `55440f476` — pair the modular path with an fp32-SiLU reference, tighten tolerance.

## Test Plan

On a gfx950 (MI355X) host, inside the ROCm test image:

```bash
pytest -q tests/kernels/moe/test_block_fp8.py::test_w8a8_block_fp8_fused_moe
```

## Test Result

Before the reference fixes (base tolerance, large-K/large-N shapes failing):

```
FAILED test_w8a8_block_fp8_fused_moe[0-dtype0-block_size0-1-2-128-4608-7168]
  Greatest absolute difference: ~0.06 (up to 0.035 allowed)
...
30 failed, 130 passed, 20 skipped
```

After this change (tight tolerance retained, faithful reference):

```
160 passed, 20 skipped, 17 warnings in 348.70s (0:05:48)
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Review by divakar-amd — COMMENTED** [review]
Hi, we had a similar issue with some other test cases too. Can you take look at the following PR and re-use some of the methods if applicable? - https://github.com/vllm-project/vllm/pull/46658

**Comment by stefankoncarevic** [comment]
> Hi, we had a similar issue with some other test cases too. Can you take look at the following PR and re-use some of the methods if applicable? - #46658

Thanks for the pointer to #46658. I looked into reusing those methods, but after measuring on gfx950 they don't fit this case.

The failure here isn't a handful of isolated 1-ULP outliers; it's fp8 block-quant accumulation noise that's **broadly distributed but small** (thousands of elements slightly over the base tolerance on the large-K/large-N shapes, hard-capped, no element exceeds 0.08, worst ~0.066). Why the #46658 tools don't transfer:

- The MoE output is **bf16, not fp8**, so `fp8_ulp_distance` doesn't apply; and bf16 ULP distance is misleading here because the output has many near-zero values (max ~30k ULP while the abs error stays ~0.05).
- An outlier-**count** bound doesn't fit either, the outlier fraction varies per shape and would need thousands allowed.
- I also measured a `rtol=1e-2` relative tolerance (the bf16 approach from that PR): even with the base atol, ~587 elements still violate on the worst shapes, so it fails the outlier bound on the 14.68M-element shape.

So a bounded **absolute** tolerance is the right tool here. I did adopt the applicable part of that PR though: the relaxation is now **ROCm-only** (CUDA stays strict), and I scoped it to just the shapes that actually fail (`K>=4096 and N>=1024`), leaving N=128 and other shapes on the strict tolerance.

**Comment by AndreasKaratzas** [comment]
> > Hi, we had a similar issue with some other test cases too. Can you take look at the following PR and re-use some of the methods if applicable? - #46658
> 
> Thanks for the pointer to #46658. I looked into reusing those methods, but after measuring on gfx950 they don't fit this case.
> 
> The failure here isn't a handful of isolated 1-ULP outliers; it's fp8 block-quant accumulation noise that's **broadly distributed but small** (thousands of elements slightly over the base tolerance on the large-K/large-N shapes, hard-capped, no element exceeds 0.08, worst ~0.066). Why the #46658 tools don't transfer:
> 
> * The MoE output is **bf16, not fp8**, so `fp8_ulp_distance` doesn't apply; and bf16 ULP distance is misleading here because the output has many near-zero values (max ~30k ULP while the abs error stays ~0.05).
> * An outlier-**count** bound doesn't fit either, the outlier fraction varies per shape and would need thousands allowed.
> * I also measured a `rtol=1e-2` relative tolerance (the bf16 approach from that PR): even with the base atol, ~587 elements still violate on the worst shapes, so it fails the outlier bound on the 14.68M-element shape.
> 
> So a bounded **absolute** tolerance is the right tool here. I did adopt the applicable part of that PR though: the relaxation is now **ROCm-only** (CUDA stays strict), and I scoped it to just the shapes that actually fail (`K>=4096 and N>=1024`), leaving N=128 and other shapes on the strict tolerance.

This sounds like a bug. Have we filed this somewhere? Is it AITER, Quark or vLLM?

**Comment by stefankoncarevic** [comment]
> > > Hi, we had a similar issue with some other test cases too. Can you take look at the following PR and re-use some of the methods if applicable? - #46658
> > 
> > 
> > Thanks for the pointer to #46658. I looked into reusing those methods, but after measuring on gfx950 they don't fit this case.
> > The failure here isn't a handful of isolated 1-ULP outliers; it's fp8 block-quant accumulation noise that's **broadly distributed but small** (thousands of elements slightly over the base tolerance on the large-K/large-N shapes, hard-capped, no element exceeds 0.08, worst ~0.066). Why the #46658 tools don't transfer:
> > 
> > * The MoE output is **bf16, not fp8**, so `fp8_ulp_distance` doesn't apply; and bf16 ULP distance is misleading here because the output has many near-zero values (max ~30k ULP while the abs error stays ~0.05).
> > * An outlier-**count** bound doesn't fit either, the outlier fraction varies per shape and would need thousands allowed.
> > * I also measured a `rtol=1e-2` relative tolerance (the bf16 approach from that PR): even with the base atol, ~587 elements still violate on the worst shapes, so it fails the outlier bound on the 14.68M-element shape.
> > 
> > So a bounded **absolute** tolerance is the right tool here. I did adopt the applicable part of that PR though: the relaxation is now **ROCm-only** (CUDA stays strict), and I scoped it to just the shapes that actually fail (`K>=4096 and N>=1024`), leaving N=128 and other shapes on the strict tolerance.
> 
> This sounds like a bug. Have we filed this somewhere? Is it AITER, Quark or vLLM?

Thanks, based on everything measured **I don't think this is a functional bug**.

## Where it lives

Pure vLLM Triton path `fused_experts` → `fused_experts_impl` (Triton), `m_out` via `TritonExperts`, and the test's own `native_per_token_group_quant_fp8` reference. No AITER, no Quark. gfx950 has no tuned fused-MoE config, so it falls back to default tiling, which shifts accumulation order vs. the platforms the 0.035 tolerance was calibrated on.

## Why I think it's expected noise, not a bug

- **Deterministic** across repeated runs at fixed seed (not flaky, not a GPU fault).
- **Present already in pure bf16** (no fp8): same kernel/shape gives max abs diff ~0.0625 — so most of it is bf16 reduction-order rounding; fp8 block-quant only adds a little.
- **Scales with the reduction length K**: K=128 → ~1e-4, K=7168 → ~0.06. N=128 shapes stay within the base tolerance.
- **Bounded and tiny in aggregate**: worst raw abs diff ~0.086 (only on E=2), and the mean relative error is ~0.014 across every shape/E.

## Methods I evaluated (incl. #46658)

Per the review pointer to #46658, I measured each candidate on the failing shapes (K=7168, N>=1024):

| Method | Source | Result |
|---|---|---|
| bf16 ULP outlier-count | #46658 (layernorm) | **FAIL on every shape** output has many near-zero values, so max ULP ~31k and ~48M/58.7M elems are ULP≠0; no budget separates good/bad |
| fp8 ULP distance | #46658 | N/A MoE output is bf16, not fp8 |
| relaxed rtol=1e-2 | #46658 | insufficient ~587 elems still violate on the worst shape |
| abs-outlier-count @0.035 | #46658-style | shape-dependent — passes E>=8, fails E=2 (thousands of small outliers) |
| **mean-relative < 0.05** | `test_triton_moe_ptpc_fp8.py` | **PASS all**, ~0.014 (≈3.5x margin), independent of E/shape |
| **bounded abs, atol=rtol=0.08** | `test_modular_kernel_combinations.py` (AITER FP8) | **PASS all**, 0 violations |

Takeaway: the ULP method comes from the layernorm/fp8 context and doesn't transfer (bf16 output + near-zero values). The two methods the **MoE** suite already uses mean-relative error and a bounded absolute tolerance both independently confirm the error is small and bounded.

## Which one for this test

- **(A) Bounded absolute tolerance (current PR):** `atol=rtol=0.08`, ROCm-only, scoped to `K>=4096 and N>=1024`. CUDA and smaller shapes stay strict. Minimal, matches the AITER-FP8 MoE test style.
- **(B) Mean-relative error `< 0.05`:** cleanest single check, uniform across all shapes/E (~0.014), no per-shape scoping needed. Already used in `test_triton_moe_ptpc_fp8.py`.
- **(C) Treat as a bug:** Open an issue.

**Comment by mawong-amd** [comment]
Not opposed to increasing tolerances per se, but I think we can pin down some of the sources of numerical divergence more carefully. I looked into this test group a while back and found the following sources of divergence on ROCm:

1. Torch reference vs Triton/HIP FP8 group quant kernel.
For a given tensor `A`, Torch calculates the scale `s` as `A.abs().max().clamp(min=eps).float() / 448.0`. Then it calculates `A / s` elementwise under the hood by first computing the inverse scale `1.0 / s`, then multiplying each element of `A` by this inverse scale.
This can be different from what Triton/HIP do. For one, they might just calculate `A / s` directly rather than taking the inverse scale. `A * (1.0 / s)` might differ from `A / s` by 2 ULP because the reciprocal/floating point division assembly instructions (`v_rcp_f32/f16/bf16` or `v_div_f32/f16/bf16`) is only accurate to 1 ULP rather than 0.5 ULP.
We can ameliorate this by using the Triton/HIP FP8 group quant functions rather than the Torch implementation in the reference.

2. Torch reference vs Triton `fused_moe` W8A8 block kernel.
The dtype handling here particularly of intermediate results is not the same. The order of accumulation in the matmuls are not necessarily the same as well. In addition, the Triton kernel fuses the `topk_weight` multiplication in the `fused_moe` kernel, whereas Torch leaves it to the penultimate intermediate tensor (and furthermore Torch performs bf16 x bf16 multiplication by silently upcasting both Tensors to fp32 and downcasting back afterward).
We can ameliorate some of this by being slightly more consistent with Triton in terms of how we structure our operations in the Torch reference.

3. I don't believe these affected the result in my analysis, but the `SiluAndMul` and `moe_sum` kernels may also have different numerical behavior compared to the Torch implementations.

There may yet be other sources of difference. It might be worth seeing if we can pin some of them down to reduce the net divergence, so that we can perhaps loosen the tolerance less or even fall back to simpler ULP based checks (e.g. all elements of some intermediate tensors below a certain ULP difference, and only a small percentage of outlier elements exceed atol/rtol).

**Comment by stefankoncarevic** [comment]
> Not opposed to increasing tolerances per se, but I think we can pin down some of the sources of numerical divergence more carefully. I looked into this test group a while back and found the following sources of divergence on ROCm:
> 
> 1. Torch reference vs Triton/HIP FP8 group quant kernel.
>    For a given tensor `A`, Torch calculates the scale `s` as `A.abs().max().clamp(min=eps).float() / 448.0`. Then it calculates `A / s` elementwise under the hood by first computing the inverse scale `1.0 / s`, then multiplying each element of `A` by this inverse scale.
>    This can be different from what Triton/HIP do. For one, they might just calculate `A / s` directly rather than taking the inverse scale. `A * (1.0 / s)` might differ from `A / s` by 2 ULP because the reciprocal/floating point division assembly instructions (`v_rcp_f32/f16/bf16` or `v_div_f32/f16/bf16`) is only accurate to 1 ULP rather than 0.5 ULP.
>    We can ameliorate this by using the Triton/HIP FP8 group quant functions rather than the Torch implementation in the reference.
> 2. Torch reference vs Triton `fused_moe` W8A8 block kernel.
>    The dtype handling here particularly of intermediate results is not the same. The order of accumulation in the matmuls are not necessarily the same as well. In addition, the Triton kernel fuses the `topk_weight` multiplication in the `fused_moe` kernel, whereas Torch leaves it to the penultimate intermediate tensor (and furthermore Torch performs bf16 x bf16 multiplication by silently upcasting both Tensors to fp32 and downcasting back afterward).
>    We can ameliorate some of this by being slightly more consistent with Triton in terms of how we structure our operations in the Torch reference.
> 3. I don't believe these affected the result in my analysis, but the `SiluAndMul` and `moe_sum` kernels may also have different numerical behavior compared to the Torch implementations.
> 
> There may yet be other sources of difference. It might be worth seeing if we can pin some of them down to reduce the net divergence, so that we can perhaps loosen the tolerance less or even fall back to simpler ULP based checks (e.g. all elements of some intermediate tensors below a certain ULP difference, and only a small percentage of outlier elements exceed atol/rtol).

Thanks for the detailed pointers I went through each source you listed on gfx950 (MI355X) rather than just widening the tolerance. Summary of what I found and changed:

**1. FP8 group-quant.** The test's reference quantized with a native torch path whose boundary rounding differs from the production `per_token_group_quant_fp8` kernel (the scale reciprocal itself is negligible, ~1e-9). ~0.06–0.10% of elements flipped to the adjacent fp8 code, and over a K=7168 reduction that accumulated. Fixed by having the reference call `per_token_group_quant_fp8` for both quant steps, so quantization is identical to the kernels (matmul stays torch/fp32).

**2a / SiLU dtype.** This turned out to be the real remaining gap: the modular path fuses SiLU+quant in fp32 (`silu_and_mul_per_block_quant`), while `fused_experts`/the reference round the intermediate SiLU to bf16 before quant. On large K/N that ~1-ULP difference pushed `m_out` past tolerance. Fixed by pairing each kernel with a reference at its own precision — `out` vs the bf16-SiLU reference, `m_out` vs an fp32-SiLU reference.

**2b / accumulation order.** After the above, the only residual is the fp8 block-matmul accumulation order (gfx950 has no tuned fused-MoE config, so default tiling). It's deterministic and bounded (~0.03, present already in pure bf16), and sits comfortably under the base tolerance.

**2c / topk_weight and moe_sum.** Both measured as non-contributors: `moe_sum` is bit-identical to `torch.sum` (0.0), and moving topk_weight to fp32 doesn't change the max error. Left as-is.

**Result.** With 1 and 2a addressed I could drop the ROCm tolerance override entirely and keep the tight base tolerance (0.035, 0.039 for M≥8192). Full `test_w8a8_block_fp8_fused_moe` sweep: **160 passed, 20 skipped**, no override.

I split this into two commits (quant alignment; paired fp32-SiLU reference + tightened tolerance).

File: tests/kernels/moe/test_block_fp8.py [hunk 1]
```diff
@@ -116,16 +124,23 @@ def torch_w8a8_block_fp8_moe(a, w1, w2, w1_s, w2_s, topk_weight, topk_ids, block
     topk_ids = topk_ids.view(-1)
 
     _, block_k = block_shape[0], block_shape[1]
-    a_q, a_s = native_per_token_group_quant_fp8(a, block_k)
+    # Quantize with the production per-token-group fp8 kernel (same HIP/CUDA op the
+    # kernels use) so the reference is bit-identical here; removes ~0.06-0.10% fp8
+    # boundary-flip divergence that otherwise accumulates over K. Matmul stays fp32.
+    a_q, a_s = per_token_group_quant_fp8(a, block_k, dtype=torch.float8_e4m3fn)
```
Comment (AndreasKaratzas) [inline tests/kernels/moe/test_block_fp8.py]: There is a current_platform.fp8_dtype() or something, did you try that one?

Comment (stefankoncarevic) [reply tests/kernels/moe/test_block_fp8.py]: Good call, done. Switched both call sites to current_platform.fp8_dtype().

**Review by mawong-amd — APPROVED** [review]

**Review by AndreasKaratzas — APPROVED** [review]
LGTM
